crates/handlers/src/admin/v1/users/set_password.rs

```diff
@@ -0,0 +1,270 @@
+    passwords::PasswordManager,
+};
+
+#[derive(Debug, thiserror::Error, OperationIo)]
+#[aide(output_with = "Json<ErrorResponse>")]
+pub enum RouteError {
+    #[error(transparent)]
+    Internal(Box<dyn std::error::Error + Send + Sync + 'static>),
+
+    #[error("Password is too weak")]
+    PasswordTooWeak,
+
+    #[error("Password hashing failed")]
+    Password(#[source] anyhow::Error),
+
+    #[error("User ID {0} not found")]
+    NotFound(Ulid),
+}
+
+impl_from_error_for_route!(mas_storage::RepositoryError);
+
+impl IntoResponse for RouteError {
+    fn into_response(self) -> axum::response::Response {
+        let error = ErrorResponse::from_error(&self);
+        let status = match self {
+            Self::Internal(_) | Self::Password(_) => StatusCode::INTERNAL_SERVER_ERROR,
+            Self::PasswordTooWeak => StatusCode::BAD_REQUEST,
+            Self::NotFound(_) => StatusCode::NOT_FOUND,
+        };
+        (status, Json(error)).into_response()
+    }
+}
+
+fn password_example() -> String {
+    "hunter2".to_owned()
+}
+
+/// # JSON payload for the `POST /api/admin/v1/users/:id/set-password` endpoint
+#[derive(Deserialize, JsonSchema)]
+#[schemars(rename = "SetUserPasswordRequest")]
+pub struct Request {
+    /// The password to set for the user
+    #[schemars(example = "password_example")]
+    password: String,
+
+    /// Skip the password complexity check
+    skip_password_check: Option<bool>,
+}
+
+pub fn doc(operation: TransformOperation) -> TransformOperation {
+    operation
+        .id("setUserPassword")
+        .summary("Set the password for a user")
+        .tag("user")
+        .response_with::<200, StatusCode, _>(|t| t.description("Password was set"))
+        .response_with::<400, RouteError, _>(|t| {
+            let response = ErrorResponse::from_error(&RouteError::PasswordTooWeak);
+            t.description("Password is too weak").example(response)
+        })
+        .response_with::<404, RouteError, _>(|t| {
+            let response = ErrorResponse::from_error(&RouteError::NotFound(Ulid::nil()));
+            t.description("User was not found").example(response)
+        })
+}
+
+#[tracing::instrument(name = "handler.admin.v1.users.set_password", skip_all, err)]
+pub async fn handler(
+    CallContext {
+        mut repo, clock, ..
+    }: CallContext,
+    NoApi(mut rng): NoApi<BoxRng>,
+    State(password_manager): State<PasswordManager>,
+    id: UlidPathParam,
+    Json(params): Json<Request>,
+) -> Result<StatusCode, RouteError> {
+    let user = repo
+        .user()
+        .lookup(*id)
+        .await?
+        .ok_or(RouteError::NotFound(*id))?;
+
+    let skip_password_check = params.skip_password_check.unwrap_or(false);
+    tracing::info!(skip_password_check, "skip_password_check");
+    if !skip_password_check
+        && !password_manager
+            .is_password_complex_enough(&params.password)
+            .unwrap_or(false)
+    {
+        return Err(RouteError::PasswordTooWeak);
```

Comment:
I think this will give a `PasswordTooWeak` error code if the password manager is disabled, which could be a bit misleading, should we have an error code just for it?

Reply:
Oh you're right! Did that in [87482b4](https://github.com/matrix-org/matrix-authentication-service/pull/3021/commits/87482b422230227dfdf30fa21fcc068578a2ad0c)